crates/red_knot_python_semantic/resources/mdtest/terminal_statements.md

```diff
@@ -0,0 +1,158 @@
+# Terminal statements
+
+## Introduction
+
+Terminal statements complicate a naive control-flow analysis.
+
+As a simple example:
+
+```py
+def f(cond: bool) -> str:
+    if cond:
+        x = "test"
+    else:
+        raise ValueError
+    return x
+
+def g(cond: bool):
+    if cond:
+        x = "test"
+        reveal_type(x)  # revealed: Literal["test"]
+    else:
+        x = "unreachable"
+        reveal_type(x)  # revealed: Literal["unreachable"]
+        raise ValueError
+    reveal_type(x)  # revealed: Literal["test"]
+```
+
+In `f`, we should be able to determine that the `else` branch ends in a terminal statement, and that
+the `return` statement can only be executed when the condition is true. We should therefore consider
+the reference always bound, even though `x` is only bound in the true branch.
+
+Similarly, in `g`, we should see that the assignment of the value `"unreachable"` can never be seen
+by the final `reveal_type`.
+
+## `return` is terminal
+
+```py
+def f(cond: bool) -> str:
+    if cond:
+        x = "test"
+    else:
+        return "early"
+    return x  # no possibly-unresolved-reference diagnostic!
+
+def g(cond: bool):
+    if cond:
+        x = "test"
+        reveal_type(x)  # revealed: Literal["test"]
+    else:
+        x = "unreachable"
+        reveal_type(x)  # revealed: Literal["unreachable"]
+        return
+    reveal_type(x)  # revealed: Literal["test"]
+```
+
+## `continue` is terminal within its loop scope
+
+```py
+def f(cond: bool) -> str:
+    while True:
+        if cond:
+            x = "test"
+        else:
+            continue
+        return x
+
+def g(cond: bool, i: int):
+    x = "before"
+    for _ in range(i):
+        if cond:
+            x = "loop"
+            reveal_type(x)  # revealed: Literal["loop"]
+        else:
+            x = "continue"
+            reveal_type(x)  # revealed: Literal["continue"]
+            continue
+        reveal_type(x)  # revealed: Literal["loop"]
+    reveal_type(x)  # revealed: Literal["before", "loop"]
```

Comment:
Ooh, this is now incorrect:

If `i <= 0`, the loop body won't execute, and `x == "before"`.
If `i > 0 and cond`, the loop body will execute `i` times, each time assigning `"loop"`.
If `i > 0 and not cond`, the loop body will execute `i` times, each time assigning `"continue"`.

The `continue` statement should mark its flow as unreachable _for when we join the `if` branches_.  But it should _not_ be considered unreachable when we join with the `for` loop's continuation and exit paths.

Reply:
After discussing in Discord, we've decided that this incorrect result is not because of the new terminal statement handling, but because we're not currently handling _any_ loop back-links (including `continue` statements) pending fixpoint support in salsa.  cf https://github.com/astral-sh/ruff/issues/14160